### DISCUSSION TITLE
Casting with 'as' should propagate an error rather than panic

### DISCUSSION THREAD

**Discussion by rcorre** [vlang/v#11873] [Ideas]
I think there are two benefits to altering `as` to return an error/option:

One is that it removes a language construct that can hide error flow. `as` looks pretty harmless, and since it is missing the usual `if`/`or`/`?` error-handling constructs, it doesn't jump out as a place your program might crash. Functions which use `as` should have to handle the error, or explicitly `or { panic() }` if a panic is truly desired.

Second is that it could help make code less verbose when you expect only a certain type. I've notice this happens a lot with `x.json2`. For example, instead of writing this:

```v
fn get_method(obj map[string]json2.Any) ?string {
	if m := obj['method'] {
		if m is string {
			return m
		} else {
			return error('invalid method')
		}
	} else {
		return error('missing method')
	}
}
```

You could remove one layer of nested `if`s:

```v
fn get_method(obj map[string]json2.Any) ?string {
	if m := obj['method'] {
		return m as string
	} else {
		return error('missing method')
	}
}
```

**Comment by dumblob** [comment]
This sounds like a good idea in general (see also https://github.com/vlang/v/issues/2030 ) though the example above is not explicit about the fact `m as string` returns an optional.

With functions one needs to add `?` suffix to propagate the error. And here for consistency and to make it **explicit** an appended `?` would also be needed.

But how to do it syntactically? Should we write `return m as? string`? Or rather `return (m as string)?`? Or both would be valid?

**Comment by rcorre** [comment]
I'd prefer `(m as string)?`, or even `m as string ?` (if operator precedence allows). I'd rather not add a new operator `as?`, since you'd still need regular `as` for `m as string or { "" }`

**Comment by dumblob** [comment]
`as?` wasn't meant to be one token. It was meant to be 2 separate tokens `as` `?` :wink:.

There is also `m? as string` (same as `m ? as string` and as `m ?as string`), but that doesn't convey the message IMHO.

**Comment by medvednikov** [comment]
I agree, this is definitely needed.

Although I'm not sure if it should be the only option, or work like maps (`x  := m[key]; y := m[key] or { ... }`), where both are allowed.

**Reply by dumblob** [reply]
Well, for maps I'm more and more inclined to the "defaultdict" approach outlined in https://github.com/vlang/v/issues/10837#issuecomment-882827508 . So actually maps are not a good example because they offer more space for a better solution (e.g. in the form of the "defaultdict").

For `as` I'd **always** propagate an error by supporting the `?` (in the future `!` as per https://github.com/vlang/rfcs/pull/7 ) syntax. On top of this, I'd turn off the propagation if inside `unsafe`.

**Comment by dumblob** [comment]
Btw. supporting both `m as? string` and `(m as string)?` can be seen as the same syntax principle as with `!(a in [0, 1, 2])` and `a !in [0, 1, 2]`.

**Comment by crthpl** [comment]
I think that `as`/`in` should be given higher precedence (move `in_as` all the way down in `token.v`)

**Comment by danieldaeschle** [comment]
Wasn't there also the idea to write `unsafe { x as string }` instead?

**Comment by dumblob** [comment]
> Wasn't there also the idea to write `unsafe { x as string }` instead?

Yes, why not (IIRC I proposed that also somewhere). But **in addition** to explicit propagation of the error if used outside of `unsafe { }` :wink: (because there were some just enough convincing arguments to not bother the user because of the ubiquitous nature of sum types).

**Comment by danieldaeschle** [comment]
It's hard to keep track of the possibilities if we add several compiler magic methods to everything :D

**Comment by dumblob** [comment]
> It's hard to keep track of the possibilities if we add several compiler magic methods to everything :D

Well, safety is hard :smiley:.
